report/5.Conclusion/Conclusion.tex

```diff
@@ -1,3 +1,75 @@
+
+The similarities with the English language is well exemplified in snippet \ref{lis:ConclusionWhile}.
+\begin{lstlisting}[language=hcl,label=lis:ConclusionWhile,caption=English-like syntax example]
+var j = 3
+func subtract = () : none { j = j - 1 }
+func jIsGreaterThan1 = () : bool { 1 lessThan j }
+subtract while jIsGreaterThan1
+\end{lstlisting}
+In snippet \ref{lis:ConclusionWhile}, lines 1 to 3 consists of declarations, line 8 consists of the function-call to While.
+The While function takes two Lambda expressions as arguments, in this case, two function-calls.
+The similarities between the function-call on line 8 and the equivalent English sentence is obvious.
+
+Since user-tests have not been conducted, it cannot be verified whether the English-like syntax allows for easier transition, from English to HCL, for novices.
+
+The project group seeks to improve the, otherwise steep, learning curve of high-order functionality.
+More precisely, HCL's handling of lambda expressions need to be intuitive.
+This was accomplished by syntactically treating any lambda expression as an anonymous implicit function definition.
+Snippet \ref{lis:ConclusionMapFilter} shows an example of how the passing of a lambda expression look in HCL.
+
+\begin{lstlisting}[language=hcl,label=lis:ConclusionMapFilter,caption=High-order functionality in HCL]
+[2, 4, 6, 8] map { value * 2 } filter { value greaterThan 10 } print
+\end{lstlisting}
+
+As mentioned, user-tests have not been conducted.
+Therefore, it cannot be verified whether the high-order functionality of HCL would be considered intuitive by novices.
+
+The programming paradigms have been simplified in HCL.
+Every operator in HCL is a function, which results in the syntax being uniform in every aspect of the language.
+This simplification removes ambiguity in the syntax when dealing with built-in and user defined functions in the same context.
+Snippet \ref{lis:ConclusionSimplified} shows an example of how this concept is handled in HCL.
+
+\begin{lstlisting}[language=hcl,label=lis:ConclusionSimplified,caption=Built-in and user defined function]
+var plus = (num left, num right): num { left + right }
+#built in addition
+var res1 = 2 + 3
+#user defined addition
+var res2 = 2 plus 3 
+\end{lstlisting}
+
+The two function calls on lines 3 and 5 both evaluates to the same value and with no ambiguity in the syntax.
+
+This finishes the conclusion of the problem-statement.
+
+\subsection{Conclusion on the solution} %Shit rhymes
+HCL is a fully fledged, Turing complete language.
+Every part of the syntax has an equivalent semantic context. 
+The compiler is capable of compiling most syntactical constructs that are valid in HCL to C++, barring any bugs and faults in the compiler as of the writing of this report.
+
+The HCL compiler is an LL(K), recursive decent, single parse, compiler.
```

Comment:
Det hedder "single pass", og nej det er den ikke. Single pass betyder, at vi læser tokens, bygger AST, typechecker og genererer kode i 1 hug. Vi har delt det op i tre -> først læser vi alle tokens (lexer), så bygger vi et helt AST fra tokens (parser og typechecker), og så genererer vi kode ud fra AST (code gen).

Udover det, skal der ikke være komma efter "parse"

Reply:
Slide 23 og 24 heri beskriver single/multi pass compiler
https://www.moodle.aau.dk/pluginfile.php/1121315/mod_label/intro/SPOF18-Exam-2.pdf

Reply:
vi har rent teknisk en "single-pass parser" idet at vi ikke behøver køre hele træet igennem for at lave typechecking efter vi har parset.
men ikke en helt single-pass compiler

Reply:
also single pass fordi vi kun går igennem den rå input kode en gang - sammenlignet med C# hvor du bliver nød til at læse det ele igennem først for at tjekke alle klasser og funktioner da de godt kan defineres efter de bliver brugt.